### PR TITLE
kill_stage_effect: Add kill and stage effects

### PR DESCRIPTION
**Pull request by SurprisedBird**
1. Add kill_effect.py
2. Add stage_effect.py
3. Fix steal_effect.py
4. Update utils.py for stage validation

### CODE REVIEW

**Review by asternotus — APPROVED** [review]
Great work!

File: effects/kill_effect.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+from citizens.player import Player
+import user_interaction
+import utils
+
+
+class KillEffect(Effect):
+    def __init__(self, name: str, game: Game, creator: Citizen):
+        Effect.__init__(self, name, game, creator)
+
+    def activate(self, target_number):
```
Comment (elias-gvin) [inline effects/kill_effect.py]: target_number should not be a parameter of activate method

File: effects/kill_effect.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        target_number = user_interaction.read_index('config.type_target')
+        while(not self.validate(target_number)):
+            user_interaction.show_active_instant("config.type_error")
+            target_number = user_interaction.read_index("config.type_target")
+
```
Comment (elias-gvin) [inline effects/kill_effect.py:18]: Code snippet repeats in several places: StageEffect, KillEffect and others.
Code duplication. Should be moved to separate function inside of utils or inside a base class (should be discussed)

File: utils.py [hunk 3]
```diff
@@ -8,4 +8,8 @@ def is_citizen_in_range(target_number: int, citizens: List[Citizen]) -> bool:
 
 
 def validate_citizen_target_number(target_number: int, citizens: List[Citizen]) -> bool:
-    return target_number is not None and is_citizen_in_range(target_number, citizens) and citizens[target_number].is_alive
+    return target_number is not None and is_citizen_in_range(target_number-1, citizens) and citizens[target_number-1].is_alive
+
+
+def stage_validate_citizen_target_number(target_number: int, citizens: List[Citizen], creator: Citizen) -> bool:
+    return target_number is not None and is_citizen_in_range(target_number-1, citizens) and citizens[target_number-1] is not creator and citizens[target_number-1].is_alive
```
Comment (elias-gvin) [inline utils.py]: This code won't be used outside of StageEffect, we don't need to move this logic to utils module (which contain commonly used code).

If you want to avoide code duplication in StageEffect - you can write:
return (utils.validate_citizen_target_number(target_number, citizens) and citizens[target_number-1] is not creator)
inside validate method of StageEffect

File: effects/steal_effect.py [hunk 4]
```diff
@@ -1,14 +1,13 @@
...
 from citizens.citizen import Citizen
-from citizens.spy import Spy
 import user_interaction
 import utils
 
 
 class StealEffect(Effect):
 
-    def __init__(self, name: str, game: Game, creator: Citizen) -> None:
+    def __init__(self, name: str, game: Game, creator: Citizen):
```
Comment (elias-gvin) [inline effects/steal_effect.py]: You should use -> None type hint for every method whitch doesn't return any value.
__init__ is not an exception

File: effects/stage_effect.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        target_number = user_interaction.read_index('config.type_target')
+        while(not self.validate(target_number)):
+            user_interaction.show_active_instant("config.type_error")
+            target_number = user_interaction.read_index("config.type_target")
+
+        self.target = self.game.citizens[target_number - 1]
+
+    def resolve(self):
+        self.creator._staging_available = False
+        # self.creator.allowed_actions.remove(CARD_USAGE)
```
Comment (elias-gvin) [inline effects/stage_effect.py]: please, use instead of CARD_USAGE -> ActionType.CARD_USAGE

File: effects/stage_effect.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            target_index = [index for index, citizen in enumerate(
+                self.game.citizens) if citizen is self.target][0]
+
+            self.game.citizens[creator_index], self.game.citizens[
+                target_index] = self.game.citizens[target_index], self.game.citizens[creator_index]
+
+            user_interaction.save_global('config.citizen_die')
+
+    def on_clear(self):
+        # self.creator.allowed_actions.append(CARD_USAGE)
```
Comment (elias-gvin) [inline effects/stage_effect.py]: please, use instead of CARD_USAGE -> ActionType.CARD_USAGE

File: effects/kill_effect.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                self.creator.stolen_cards.append(self.target.citizen_card)
+                self.target.citizen_card = None
+                user_interaction.save_active('config.card_stolen')
+            else:
+                user_interaction.save_active('config.no_card')
+
+    def on_clear(self):
+        pass
+
+    def validate(self, target_number):
```
Comment (elias-gvin) [inline effects/kill_effect.py]: Please, use type hints for method arguments and method return values

def validate(self, target_number: int) -> None:

Rewrite other methods in this style.

File: effects/kill_effect.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            target_number = user_interaction.read_index("config.type_target")
+
+        self.target = self.game.citizens[target_number - 1]
+
+    def resolve(self):
+
+        if isinstance(self.target, Player):
+            self.target.hp -= 1
+            user_interaction.save_passive('config.lose_hp')
+            user_interaction.save_global('config.failed_kill')
```
Comment (elias-gvin) [inline effects/kill_effect.py:26]: What if player was killed?

File: effects/kill_effect.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+        else:
+            self.target.hp -= 1
+            user_interaction.save_global('config.citizen_die')
```
Comment (elias-gvin) [inline effects/kill_effect.py:30]: What if citizen - Spy?

File: effects/stage_effect.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            if self.creator.citizen_card is not None:
+                self.creator.stolen_cards.append(self.creator.citizen_card)
+                self.creator.citizen_card = None
+
+            self.creator.citizen_card = self.target.citizen_card
+
+            creator_index = [index for index, citizen in enumerate(
+                self.game.citizens) if citizen is self.creator][0]
+            target_index = [index for index, citizen in enumerate(
+                self.game.citizens) if citizen is self.target][0]
```
Comment (elias-gvin) [inline effects/stage_effect.py]: We should move this logic to utils.
Something like
utils.find_citizen_index(target_citizen: Citizen, citizens: List[Citizen])

**Review by elias-gvin — CHANGES_REQUESTED** [review]

File: effects/steal_effect.py [hunk 11]
```diff
@@ -1,14 +1,13 @@
...
         Effect.__init__(self, name, game, creator)
 
     def activate(self):
```
Comment (elias-gvin) [inline effects/steal_effect.py]: Inside of all activate methods we should change effect status to "EffectStatus.ACTIVATE".
And inside all resolve methods we should change effect status to "EffectStatus.FINISHED"
!!!
important. Should be discussed.
Global change

File: effects/kill_effect.py (hunk 8 above)
Comment (elias-gvin) [reply effects/kill_effect.py:26]: I think, there is a solution. We can write message like this in global output: 'Было совершено нападение на _имя_  жителя' - without specifying if he died or not.

**Comment by elias-gvin** [comment]
Great work. General logic of effects seems legit, but several changes are needed.
Also, this pull request lighted up some problems which were missed during previous review with @asternotus 
Should be discussed

File: effects/kill_effect.py (hunk 8 above)
Comment (SurprisedBird) [reply effects/kill_effect.py:26]: We can announce about player's death in "check win" part.

File: effects/kill_effect.py (hunk 1 above)
Comment (SurprisedBird) [reply effects/kill_effect.py]: Done

File: utils.py (hunk 3 above)
Comment (SurprisedBird) [reply utils.py]: Done

File: effects/steal_effect.py (hunk 4 above)
Comment (SurprisedBird) [reply effects/steal_effect.py]: Done

File: effects/stage_effect.py (hunk 5 above)
Comment (SurprisedBird) [reply effects/stage_effect.py]: Done

File: effects/stage_effect.py (hunk 6 above)
Comment (SurprisedBird) [reply effects/stage_effect.py]: Done

File: effects/kill_effect.py (hunk 7 above)
Comment (SurprisedBird) [reply effects/kill_effect.py]: Done

File: effects/stage_effect.py (hunk 10 above)
Comment (SurprisedBird) [reply effects/stage_effect.py]: Done

**Pull request closed without merging**